src/lib/signature_lib/schnorr.ml

```diff
@@ -509,83 +505,164 @@ module Message = struct
 
-  let make_hash ~init t ~public_key ~r =
-    let input =
-      let px, py = Inner_curve.to_affine_exn public_key in
-      Random_oracle.Input.append t
-        { field_elements = [| px; py; r |]; bitstrings = [||] }
-    in
-    let open Random_oracle in
-    hash ~init (pack_input input)
-    |> Digest.to_bits ~length:Field.size_in_bits
-    |> Inner_curve.Scalar.of_bits
+    let make_hash ~init t ~public_key ~r =
+      let input =
+        let px, py = Inner_curve.to_affine_exn public_key in
+        Random_oracle.Input.Legacy.append t
+          { field_elements = [| px; py; r |]; bitstrings = [||] }
+      in
+      let open Random_oracle.Legacy in
+      hash ~init (pack_input input)
+      |> Digest.to_bits ~length:Field.size_in_bits
+      |> Inner_curve.Scalar.of_bits
 
-  let hash = make_hash ~init:Hash_prefix_states.signature
+    let hash = make_hash ~init:Hash_prefix_states.signature
 
-  let hash_for_mainnet =
-    make_hash ~init:Hash_prefix_states.signature_for_mainnet
+    let hash_for_mainnet =
+      make_hash ~init:Hash_prefix_states.signature_for_mainnet
 
-  let hash_for_testnet =
-    make_hash ~init:Hash_prefix_states.signature_for_testnet
+    let hash_for_testnet =
+      make_hash ~init:Hash_prefix_states.signature_for_testnet
 
-  [%%ifdef consensus_mechanism]
+    [%%ifdef consensus_mechanism]
 
-  type var = (Field.Var.t, Boolean.var) Random_oracle.Input.t
+    type var = (Field.Var.t, Boolean.var) Random_oracle.Input.Legacy.t
 
-  let%snarkydef hash_checked t ~public_key ~r =
-    let input =
-      let px, py = public_key in
-      Random_oracle.Input.append t
-        { field_elements = [| px; py; r |]; bitstrings = [||] }
-    in
-    make_checked (fun () ->
-        let open Random_oracle.Checked in
-        hash ~init:Hash_prefix_states.signature (pack_input input)
-        |> Digest.to_bits ~length:Field.size_in_bits
-        |> Bitstring_lib.Bitstring.Lsb_first.of_list)
+    let%snarkydef hash_checked t ~public_key ~r =
+      let input =
+        let px, py = public_key in
+        Random_oracle.Input.Legacy.append t
+          { field_elements = [| px; py; r |]; bitstrings = [||] }
+      in
+      make_checked (fun () ->
+          let open Random_oracle.Legacy.Checked in
+          hash ~init:Hash_prefix_states.signature (pack_input input)
+          |> Digest.to_bits ~length:Field.size_in_bits
+          |> Bitstring_lib.Bitstring.Lsb_first.of_list)
 
-  [%%endif]
+    [%%endif]
+  end
+
+  module Chunked = struct
+    open Tick
+
+    type t = Field.t Random_oracle.Input.Chunked.t [@@deriving sexp]
+
+    let make_derive ~network_id t ~private_key ~public_key =
+      let input =
+        let x, y = Tick.Inner_curve.to_affine_exn public_key in
+        let id =
+          Fold_lib.Fold.(to_list (string_bits (String.of_char network_id)))
+        in
+        Random_oracle.Input.Chunked.append t
+          { field_elements =
+              [| x; y; Field.project (Tock.Field.unpack private_key) |]
```

Comment:
Why does privatekey bits get into `field_elements` but `network_id` must go into `packeds`?

Reply:
The `private_key` uses the same number of bits as the full field element, so it's more efficient to pack it explicitly here than to defer that to `Random_oracle_input`.

Reply:
Does `Field.project` do something to safely handle the change of field (i.e. scalar field to base field)?  I.e. make sure we don't exceed base field modulus..

Reply:
I was assuming it's 'not ideal, but good enough', because we're only using it as a stable-per-message random seed for the schnorr signature. I can switch to using the bits directly if that would make you more comfortable.

Reply:
Actually, my bad. This is part of deriving the nonce and so it actually gets converted to bytes and fed to blake2 (rather than field elements to poseidon) so we don't need to worry about any subsequent change of field during conversion to base field element in this case.